src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -142,82 +131,74 @@ public void Dispose()
             _logger.LogTrace("Beginning receive loop");
             try
             {
-                while (await _connection.Input.WaitToReadAsync(cancellationToken))
+                ReceiveData receiveData = new ReceiveData();
+                while (await _connection.ReceiveAsync(receiveData, cancellationToken))
                 {
-                    Message incomingMessage;
-                    while (_connection.Input.TryRead(out incomingMessage))
-                    {
-
-                        InvocationMessage message;
-                        using (incomingMessage)
-                        {
-                            message = await _adapter.ReadMessageAsync(
-                                 new MemoryStream(incomingMessage.Payload.Buffer.ToArray()), _binder, cancellationToken);
-                        }
+                    InvocationMessage message 
+                        = await _adapter.ReadMessageAsync(new MemoryStream(receiveData.Data), _binder, cancellationToken);
 
-                        var invocationDescriptor = message as InvocationDescriptor;
-                        if (invocationDescriptor != null)
-                        {
-                            DispatchInvocation(invocationDescriptor, cancellationToken);
-                        }
-                        else
+                    if (message is InvocationDescriptor invocationDescriptor)
```

Comment:
as + null is to avoid double cast. This is C#7 and there is no double cast here.